trio/_core/_gen_exports.py

```diff
@@ -0,0 +1,147 @@
+import ast
+import astor
+import os
+import textwrap
+
+SOURCE_TREE = './trio/_core'
+
+
+def get_public_methods(tree):
+    """ Return a list of methods marked as public.
+    The function walks the given tree and extracts
+    all objects that are functions and have a
+    doc string that starts with PUBLIC
+    """
+    methods = []
+    for node in ast.walk(tree):
+        for child in ast.iter_child_nodes(node):
+            if (isinstance(child, ast.FunctionDef) or
+               isinstance(child, ast.AsyncFunctionDef)) and \
+               get_doc_string(child).startswith('PUBLIC '):
```

Comment:
I still think we're going to want to use the decorator to mark functions (`any(isinstance(dec, ast.Name) and dec.id == "_public") for dec in child.decorator_list`), instead of the docstring.

Reply:
Well of course if you want decorators I will put them in again. One last remark on this. On the last pycon in Germany I talked to a peer and he took his working code and ported it from asyncio to trio. It went not smoothly but after figuring out what to do it worked.
However the performance was 30 % slower than on asyncio. Of course trio does a lot more than asyncio but that comes at a price. The impact of an empty decorator is not much but the functions you are exporting might be called a couple of thousand times a second and then even a few hundred nano seconds per call are up to a measurable millisecond.
I realized this when I tried to solve the problem from within the decorator, and my tests started to take longer.
Another suggestion to solve this is to just create a list where every name that should be exported is added. The decorator made perfectly sense when you did things at run time but now when stuff is done even before import you shouldn't use a decorator anymore.
As I write this the list is starting to get more and more my favorite as I do not even have to scan for public functions anymore I get them for free.

Reply:
I'll go for the decorator. I still don't know how python does it, but I made some tests and there really was no difference. I have to admit the decorator looks much cleaner. You learn something every day ;-)

Reply:
Remember that decorators don't really have any magic... when you see:

```python
@_public
def fn(...):
    ...
```

that's just a shorthand for 

```python
def fn(...):
    ...
fn = _public(fn)
```

So if `_public` is a simple no-op like in our example above, then this becomes:

```python
def fn(...):
    ...
fn = fn
```

So it makes sense that a no-op decorator doesn't have any effect, because it literally has no effect at all after the module finishes importing.